File: paper/scripts/rdkitjs_alert_parity.py
#!/usr/bin/env python3
"""Compare QED structural-alert counts between RDKit.js and RDKit Python.

The application computes QED's alert term in the browser (countQEDAlerts in
src/utils/chem.ts), against the 116 patterns in src/data/qedAlerts.ts, using a
WebAssembly build of RDKit that is a year older than the Python one the analysis
scripts use. The response letter states that the two agree; that statement is
only worth anything if it can be re-run, so it is re-run here.

The Python side is the reference implementation, rdkit.Chem.QED.StructuralAlerts
with HasSubstructMatch. The browser side is driven by rdkitjs_alert_parity.mjs,
which loads the same bundle the end-to-end tests serve to the browser and
mirrors countQEDAlerts pattern for pattern.

Exits non-zero on any per-molecule disagreement, on any pattern the browser
build will not compile, and on any divergence between the application's alert
patterns and RDKit's own. The last two would each let counts agree for the wrong
reason, and agreement reached that way is not evidence of anything.

Requires Node and the RDKit.js bundle in e2e/fixtures (`node
e2e/fetch-fixtures.mjs`). Neither is present in a Python-only checkout, so the
script skips explicitly rather than failing or, worse, passing quietly.

Usage:  python3 paper/scripts/rdkitjs_alert_parity.py
"""
from __future__ import annotations

import csv
import json
import pathlib
import random
import shutil
import subprocess
import sys
import tempfile

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from _paths import FDA_CSV, RESULTS, ROOT, SCRIPTS, rel as _rel  # noqa: E402

import rdkit  # noqa: E402
from rdkit import Chem, RDLogger  # noqa: E402
from rdkit.Chem import QED  # noqa: E402

RDLogger.DisableLog("rdApp.*")

N_COMPOUNDS = 500
# A fixed random subset rather than the first 500 rows: the reference set is
# ordered by ChEMBL identifier, so its head is not a representative sample and
# agreement over it would be a soft spot in the check.
SEED = 20260727

HELPER = SCRIPTS / "rdkitjs_alert_parity.mjs"
FIXTURES = ROOT / "e2e" / "fixtures"
OUT = RESULTS / "rdkitjs_alert_parity.json"


def skip(reason: str) -> int:
    """Report an unrunnable check as a skip and write nothing.

    A result file saying "did not run" deposited over one saying "500 of 500
    agreed" would be a silent downgrade, and a skip that exits 0 without saying
    so is indistinguishable from a pass.
    """
    print(f"SKIP: {reason}")
    print("      RDKit.js was not compared against RDKit Python.")
    return 0


def run_rdkitjs(smiles: list[str]) -> dict:
    with tempfile.TemporaryDirectory() as tmp:
        src = pathlib.Path(tmp) / "smiles.json"
        dst = pathlib.Path(tmp) / "counts.json"
        src.write_text(json.dumps({"smiles": smiles}), encoding="utf-8")
        subprocess.run(["node", str(HELPER), str(src), str(dst)], check=True)
        return json.loads(dst.read_text(encoding="utf-8"))


def main() -> int:
    if shutil.which("node") is None:
        return skip("node is not on PATH")
    missing = [n for n in ("RDKit_minimal.js", "RDKit_minimal.wasm")
               if not (FIXTURES / n).exists()]
    if missing:
        return skip(f"{', '.join(missing)} absent from {_rel(FIXTURES)}; "
                    "run `node e2e/fetch-fixtures.mjs`")

    population = [r["smiles"] for r in csv.DictReader(FDA_CSV.open())
                  if Chem.MolFromSmiles(r["smiles"]) is not None]
    if len(population) < N_COMPOUNDS:
        print(f"only {len(population)} parseable compounds in {_rel(FDA_CSV)}")
        return 1
    sample = random.Random(SEED).sample(population, N_COMPOUNDS)

    py_counts = [sum(1 for a in QED.StructuralAlerts if mol.HasSubstructMatch(a))
                 for mol in (Chem.MolFromSmiles(s) for s in sample)]

    js = run_rdkitjs(sample)
    js_counts = js["counts"]
    # Guarded rather than assumed: zip would quietly compare the shorter list
    # and the run would report 500 comparisons it had not made.
    if len(js_counts) != len(sample):
        print(f"RDKit.js returned {len(js_counts)} counts for {len(sample)} compounds")
        return 1

    identical = list(js["smarts"]) == list(QED.StructuralAlertSmarts)
    compiled = js["n_compiled"] == len(js["smarts"])
    disagree = [{"smiles": s, "rdkit_python": p, "rdkit_js": j}
                for s, p, j in zip(sample, py_counts, js_counts) if p != j]

    print(f"compounds: {len(sample)} of {len(population)} parseable, "
          f"seed {SEED}, from {_rel(FDA_CSV)}")
    print(f"engines: RDKit Python {rdkit.__version__} vs RDKit.js {js['rdkitjs_version']}")
    print(f"alert patterns: {len(js['smarts'])}, "
          f"{'identical to' if identical else 'DIFFERENT FROM'} "
          "rdkit.Chem.QED.StructuralAlertSmarts, "
          f"{js['n_compiled']} compiled by RDKit.js")
    print(f"  agree     {len(sample) - len(disagree):4d} / {len(sample)}")
    print(f"  disagree  {len(disagree):4d}")
    # Stated so agreement cannot be read as impressive when it is vacuous: two
    # engines that both find nothing agree perfectly.
    print(f"  alerts found: {sum(py_counts)} across "
          f"{sum(1 for c in py_counts if c)} molecules, max {max(py_counts)} in one")

    OUT.write_text(json.dumps({
        "n_compounds": len(sample),
        "n_agree": len(sample) - len(disagree),
        "n_disagree": len(disagree),
        "seed": SEED,
        "source": _rel(FDA_CSV),
        "rdkit_python_version": rdkit.__version__,
        "rdkit_js_version": js["rdkitjs_version"],
        "n_alert_smarts": len(js["smarts"]),
        "n_alert_smarts_compiled_by_rdkit_js": js["n_compiled"],
        "smarts_identical": identical,
        "alerts_total": sum(py_counts),
        "molecules_with_alert": sum(1 for c in py_counts if c),
        "disagreements": disagree,
    }, indent=2) + "\n", encoding="utf-8")
    print(f"\nwrote {_rel(OUT)}")

    if disagree:
        print(f"\nDISAGREEMENT on {len(disagree)} compound(s) "
              "(rdkit_js null means RDKit.js could not parse it):")
        for d in disagree:
            print(f"  {d['smiles']}\n    python {d['rdkit_python']}   "
                  f"rdkitjs {d['rdkit_js']}")
    if not identical:
        print("\nsrc/data/qedAlerts.ts no longer matches rdkit.Chem.QED; "
              "regenerate it with build_qed_alerts.py.")
    if not compiled:
        print(f"\nRDKit.js compiled only {js['n_compiled']} of {len(js['smarts'])} "
              "patterns, so the browser cannot see every alert.")
    return 1 if (disagree or not identical or not compiled) else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: paper/scripts/rdkitjs_alert_parity.mjs
// Count QED structural alerts with the browser build of RDKit, exactly the way
// the application does, and write the counts where the Python side can read
// them.
//
//   node paper/scripts/rdkitjs_alert_parity.mjs <input.json> <output.json>
//
// Input:  {"smiles": [...]}
// Output: {"rdkitjs_version", "smarts", "n_compiled", "counts"}
// A count is null where RDKit.js could not parse the molecule.
//
// Driven by paper/scripts/rdkitjs_alert_parity.py rather than run directly: the
// comparison is only meaningful against the Python reference computed there.
import { copyFileSync, existsSync, mkdtempSync, readFileSync, writeFileSync } from 'node:fs';
import { createRequire } from 'node:module';
import { tmpdir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..', '..');
const FIXTURES = join(ROOT, 'e2e', 'fixtures');
const ALERTS_TS = join(ROOT, 'src', 'data', 'qedAlerts.ts');

const [inPath, outPath] = process.argv.slice(2);
if (!inPath || !outPath) {
  console.error('usage: node rdkitjs_alert_parity.mjs <input.json> <output.json>');
  process.exit(2);
}

/** The alert patterns the application itself uses, read from its data module.
 *  Re-listing them here would compare RDKit.js against a copy rather than
 *  against what ships in the browser. */
function alertSmarts() {
  const src = readFileSync(ALERTS_TS, 'utf8');
  // Terminated on `];` rather than the first `]`, which SMARTS atom
  // expressions are full of.
  const block = src.match(/QED_ALERT_SMARTS[^=]*=\s*\[([\s\S]*?)\]\s*;/);
  const out = block ? [...block[1].matchAll(/'([^']*)'/g)].map(m => m[1]) : [];
  if (out.length === 0) throw new Error(`no SMARTS found in ${ALERTS_TS}`);
  return out;
}

/** RDKit.js is a CommonJS UMD bundle and this package is ESM, so Node would
 *  parse it as a module and hand back an empty object. A .cjs copy forces the
 *  CommonJS loader; locateFile keeps the 7 MB WebAssembly file where it is. */
async function loadRDKit() {
  const js = join(FIXTURES, 'RDKit_minimal.js');
  const wasm = join(FIXTURES, 'RDKit_minimal.wasm');
  if (!existsSync(js) || !existsSync(wasm)) {
    throw new Error(`RDKit.js fixtures missing from ${FIXTURES}; run node e2e/fetch-fixtures.mjs`);
  }
  const cjs = join(mkdtempSync(join(tmpdir(), 'rdkitjs-')), 'RDKit_minimal.cjs');
  copyFileSync(js, cjs);
  const initRDKitModule = createRequire(cjs)(cjs);
  return initRDKitModule({ locateFile: () => wasm });
}

// Mirrors countQEDAlerts in src/utils/chem.ts. The receiver is the molecule and
// the argument is the query: the other way round silently matches nothing.
function countAlerts(mol, queries) {
  let n = 0;
  for (const q of queries) {
    if (!q) continue;
    try {
      const hit = mol.get_substruct_match(q);
      if (hit && hit !== '{}' && hit !== '') n++;
    } catch { /* pattern not applicable to this molecule */ }
  }
  return n;
}

const { smiles } = JSON.parse(readFileSync(inPath, 'utf8'));
const smarts = alertSmarts();
const RDKit = await loadRDKit();

// A pattern the browser build will not compile is skipped rather than fatal,
// as in the application; the count of those that did compile is reported so
// that silent under-counting cannot pass as agreement.
const queries = smarts.map(s => {
  try {
    const q = RDKit.get_qmol(s);
    if (q && !q.is_valid()) { q.delete(); return null; }
    return q;
  } catch { return null; }
});

const counts = smiles.map(s => {
  // get_mol throws on some malformed input and returns an invalid molecule on
  // the rest; the application treats both as unparseable, so this does too.
  let mol = null;
  try {
    mol = RDKit.get_mol(s);
    if (!mol || !mol.is_valid()) return null;
    return countAlerts(mol, queries);
  } catch { return null; } finally { mol?.delete(); }
});

for (const q of queries) q?.delete();

writeFileSync(outPath, JSON.stringify({
  rdkitjs_version: RDKit.version(),
  smarts,
  n_compiled: queries.filter(Boolean).length,
  counts,
}));
